larpandoracontent/LArShowerRefinement/ConnectionPathwayFeatureTool.cc

```diff
@@ -0,0 +1,1187 @@
+
+void InitialRegionFeatureTool::Run(LArMvaHelper::MvaFeatureMap &featureMap, StringVector &featureOrder, const std::string &featureToolName,
+    const Algorithm *const pAlgorithm, const ParticleFlowObject *const pShowerPfo, const CartesianVector &nuVertex3D,
+    const ProtoShowerMatch &protoShowerMatch, const CartesianPointVector &showerStarts3D)
+{
+    LArMvaHelper::MvaFeatureVector toolFeatureVec;
+    this->Run(toolFeatureVec, pAlgorithm, pShowerPfo, nuVertex3D, protoShowerMatch, showerStarts3D);
+
+    if (featureMap.find(featureToolName + "_initialGapSize") != featureMap.end())
+    {
+        std::cout << "Already wrote initialGapSize feature into map! Not writing again." << std::endl;
+        throw pandora::StatusCodeException(pandora::STATUS_CODE_INVALID_PARAMETER);
+    }
+
+    featureOrder.push_back(featureToolName + "_initialGapSize");
+    featureMap[featureToolName + "_initialGapSize"] = toolFeatureVec[0];
+
+    if (featureMap.find(featureToolName + "_largestGapSize") != featureMap.end())
+    {
+        std::cout << "Already wrote largestGapSize feature into map! Not writing again." << std::endl;
+        throw pandora::StatusCodeException(pandora::STATUS_CODE_INVALID_PARAMETER);
+    }
+
+    featureOrder.push_back(featureToolName + "_largestGapSize");
+    featureMap[featureToolName + "_largestGapSize"] = toolFeatureVec[1];
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void InitialRegionFeatureTool::GetViewInitialRegionVariables(const Algorithm *const pAlgorithm, const CartesianVector &nuVertex3D,
+    const ProtoShowerMatch &protoShowerMatch, const HitType hitType, float &initialGapSize, float &largestGapSize) const
+{
+    const ProtoShower &protoShower(hitType == TPC_VIEW_U ? protoShowerMatch.m_protoShowerU
+                                                         : (hitType == TPC_VIEW_V ? protoShowerMatch.m_protoShowerV : protoShowerMatch.m_protoShowerW));
+    const CartesianVector nuVertex2D(LArGeometryHelper::ProjectPosition(pAlgorithm->GetPandora(), nuVertex3D, hitType));
+    const CartesianVector &startDirection(protoShower.m_connectionPathway.m_startDirection);
+
+    // Initial gap size
+    CaloHitVector spineHitVector(protoShower.m_spineHitList.begin(), protoShower.m_spineHitList.end());
+
+    std::sort(spineHitVector.begin(), spineHitVector.end(), LArConnectionPathwayHelper::SortByDistanceToPoint(nuVertex2D));
+    initialGapSize = (nuVertex2D - spineHitVector.front()->GetPositionVector()).GetMagnitude();
+
+    // Largest Gap Size
+    largestGapSize = m_defaultFloat;
+
+    std::vector<float> longitudinalProjections;
+
+    for (const CaloHit *const pCaloHit : protoShower.m_spineHitList)
+        longitudinalProjections.push_back(startDirection.GetDotProduct(pCaloHit->GetPositionVector() - nuVertex2D));
+
+    std::sort(longitudinalProjections.begin(), longitudinalProjections.end());
```

Comment:
discussed on slack

`I can't use the spineHitVector because they're ordered differently.. one is the sqrt(deltax^2 + deltay^2 + deltaz^2) distance and the other is a projection onto an axis.. so they're slightly different. My BDT is more powerful using these choices`